MapboxNavigation/VanishingRouteLine.swift

```diff
@@ -0,0 +1,372 @@
+        let traveledIndex = granularDistances.distanceArray[index]
+        let upcomingPoint = traveledIndex.point
+        
+        /**
+         Take the remaining distance from the upcoming point on the route and extends it by the exact position of the puck.
+         */
+        let remainingDistance = traveledIndex.distanceRemaining + calculateDistance(coordinate1: upcomingPoint, coordinate2: coordinate)
+        
+        /**
+         Calculate the percentage of the route traveled.
+         */
+        if granularDistances.distance >= remainingDistance {
+            let offSet = (1.0 - remainingDistance / granularDistances.distance)
+            if offSet >= 0 {
+                preFractionTraveled = fractionTraveled
+                fractionTraveled = offSet
+            }
+        }
+    }
+    
+    /**
+     Updates the route style layer and its casing style layer to gradually disappear as the user location puck travels along the displayed route.
+     
+     - parameter routeProgress: Current route progress.
+     */
+    public func updateRoute(_ routeProgress: RouteProgress) {
+        guard let mainRouteLayerIdentifier = identifier(routes?.first, identifierType: .route),
+              let mainRouteCasingLayerIdentifier = identifier(routes?.first, identifierType: .routeCasing) else { return }
+        
+        guard let mainRouteLayer = style?.layer(withIdentifier: mainRouteLayerIdentifier) as? MGLLineStyleLayer,
+              let mainRouteCasingLayer = style?.layer(withIdentifier: mainRouteCasingLayerIdentifier) as? MGLLineStyleLayer else { return }
+        
+        if fractionTraveled >= 1.0 {
+            // In case if route was fully travelled - remove main route and its casing.
+            style?.remove([mainRouteLayer, mainRouteCasingLayer])
+            return
+        }
+        
+        timer?.invalidate()
+        timer = nil
+        let traveledDifference = fractionTraveled - preFractionTraveled
+        if traveledDifference == 0.0 {
+            return
+        }
+        let startDate = Date()
+        timer = Timer.scheduledTimer(withTimeInterval: 0.05, repeats: true, block: { timer in
+            let timePassedInMilliseconds = Date().timeIntervalSince(startDate) * 1000
+            if timePassedInMilliseconds >= 980 {
+                timer.invalidate()
+                return
+            }
+            let newFractionTraveled = self.preFractionTraveled + traveledDifference * timePassedInMilliseconds.truncatingRemainder(dividingBy: 1000) / 1000
+            guard let mainRouteLayerGradient = self.routeLineGradient(routeProgress.route, fractionTraveled: newFractionTraveled) else { return }
+            let mainCasingLayerStops = self.routeCasingGradient(newFractionTraveled)
+            
+            mainRouteLayer.lineGradient = mainRouteLayerGradient
+            mainRouteCasingLayer.lineGradient = mainCasingLayerStops
```

Comment:
Can you make sure that this part doesn't cause any side effect? E.g. what if active guidance navigation stops suddenly on high speeds? Will it cause any issues to route line gradient?

Reply:
I'll work on this problem and check how it works.